prow/plank/reconciler.go

```diff
@@ -0,0 +1,645 @@
+				Name:      toCancel.Name,
+			},
+		}
+		if err := client.Delete(r.ctx, podToDelete); err != nil {
+			return fmt.Errorf("failed to delete pod %s/%s: %w", podToDelete.Namespace, podToDelete.Name, err)
+		}
+		return nil
+	})
+}
+
+func (r *reconciler) syncPendingJob(pj *prowv1.ProwJob) error {
+	prevPJ := pj.DeepCopy()
+
+	pod, podExists, err := r.pod(pj)
+	if err != nil {
+		return err
+	}
+
+	if !podExists {
+		// Pod is missing. This can happen in case the previous pod was deleted manually or by
+		// a rescheduler. Start a new pod.
+		id, pn, err := r.startPod(pj)
+		if err != nil {
+			if !isRequestError(err) {
+				return fmt.Errorf("error starting pod %s: %v", pod.Name, err)
+			}
+			pj.Status.State = prowv1.ErrorState
+			pj.SetComplete()
+			pj.Status.Description = "Job cannot be processed."
+			r.log.WithFields(pjutil.ProwJobFields(pj)).WithError(err).Warning("Unprocessable pod.")
+		} else {
+			pj.Status.BuildID = id
+			pj.Status.PodName = pn
+			r.log.WithFields(pjutil.ProwJobFields(pj)).Info("Pod is missing, starting a new pod")
+		}
+	} else {
+
+		switch pod.Status.Phase {
+		case corev1.PodUnknown:
+			// Pod is in Unknown state. This can happen if there is a problem with
+			// the node. Delete the old pod, we'll start a new one next loop.
+			r.log.WithFields(pjutil.ProwJobFields(pj)).Info("Pod is in unknown state, deleting & restarting pod")
+			client, ok := r.buildClients[pj.ClusterAlias()]
+			if !ok {
+				return fmt.Errorf("unknown pod %s: unknown cluster alias %q", pod.Name, pj.ClusterAlias())
+			}
+
+			r.log.WithField("name", pj.ObjectMeta.Name).Debug("Delete Pod.")
+			return client.Delete(r.ctx, pod)
+
+		case corev1.PodSucceeded:
+			// Pod succeeded. Update ProwJob, talk to GitHub, and start next jobs.
+			pj.SetComplete()
+			pj.Status.State = prowv1.SuccessState
+			pj.Status.Description = "Job succeeded."
+
+		case corev1.PodFailed:
+			if pod.Status.Reason == Evicted {
+				// Pod was evicted.
+				if pj.Spec.ErrorOnEviction {
+					// ErrorOnEviction is enabled, complete the PJ and mark it as errored.
+					pj.SetComplete()
+					pj.Status.State = prowv1.ErrorState
+					pj.Status.Description = "Job pod was evicted by the cluster."
+					break
+				}
+				// ErrorOnEviction is disabled. Delete the pod now and recreate it in
+				// the next resync.
+				client, ok := r.buildClients[pj.ClusterAlias()]
+				if !ok {
+					return fmt.Errorf("evicted pod %s: unknown cluster alias %q", pod.Name, pj.ClusterAlias())
+				}
+				r.log.WithField("name", pj.ObjectMeta.Name).Debug("Delete Pod.")
+				return client.Delete(r.ctx, pod)
+			}
+			// Pod failed. Update ProwJob, talk to GitHub.
+			pj.SetComplete()
+			pj.Status.State = prowv1.FailureState
+			pj.Status.Description = "Job failed."
+
+		case corev1.PodPending:
+			maxPodPending := r.config().Plank.PodPendingTimeout.Duration
+			maxPodUnscheduled := r.config().Plank.PodUnscheduledTimeout.Duration
+			if pod.Status.StartTime.IsZero() {
```

Comment:
So we found that initcontainers running did not cause the start time to increment ... either this or the pending timeout is bugged as a result

Reply:
This is copied from the existing code so either its broken in both implementations or working in both. Are you okay with me opening an issue for this so we can discuss and fix this seperatly?

Reply:
https://github.com/kubernetes/test-infra/issues/17688